suite2p_preprocessing: conda env switching

milesAraya opened the pull request:
### Content

Separation of https://github.com/arayabrain/optinist-for-server/pull/459 into pull requests for conda env switching scripts and suite2p related scripts. 

### Testcase
create proc file with format:
```
command: regist
roi_method: caiman
```
And run batch
`python run_expdb_batch.py -o 1`


### Known issue

> 2025-11-04 11:25:01,514 INFO: __stopwatch_callback():159 - processing done. [generate_plots()][elapsed_time: 8.881158 sec]
> 2025-11-04 11:25:01,514 INFO: generate_plots_spatial():789 - process 'generate_pca_analysis_plots' start.
> 2025-11-04 11:25:01,517 INFO: __stopwatch_callback():159 - processing done. [load_raw_timecourse_data()][elapsed_time: 0.001842 sec]
> 2025-11-04 11:25:01,530 ERROR: process_single_dataset():589 - <class 'ModuleNotFoundError'>: No module named 'studio.app.optinist.wrappers.expdb.roi_utils'

roi_utils will be added in a separate pull request. The conda env switching still works correctly.

milesAraya committed:
Updates related to conda env switching copied from feature/preprocessing_suite2p

milesAraya committed:
Merge branch 'develop-suite2p-and-batch-gui' into feature/conda-env-switch

milesAraya committed:
Additional file for better error detection

itutu-tienday commented on studio/app/common/core/experiment/experiment_reader.py:
```diff
@@ -89,9 +89,11 @@ def validate_experiment_config(cls, config: ExptConfig) -> bool:
-            assert (
-                config_dict.get(field) is not None
-            ), f"ExptConfig.{field} is required."
+            if config_dict.get(field) is None:
```
What was the need for this code update?

milesAraya replied:
I think it was convenient to have `return False` when `config_dict.get(field) is None`.

itutu-tienday replied:
It seems like it would be better to address this in barebone-stduio (the fork).
If you're okay with that, I'll handle it here (I'll roll back the commit here for now).

milesAraya replied:
Can this be considered resolved?

itutu-tienday commented on studio/config/docker/Dockerfile.dev:
```diff
@@ -100,12 +107,22 @@ RUN chown -R optinist /usr/local/lib/python3.9/site-packages/pynwb
-# pip install to conda environment for batch processing
+# pip install to all conda environments for batch processing
 RUN . ~/.bashrc && \
     conda activate expdb_batch && \
```
Is there any need to keep this conda env (expdb_batch)?
If there is little need to keep it due to the addition of expdb_batch_(caiman|suite2p), it seems appropriate to delete it to make the environment setup lighter.

milesAraya replied:
expdb_batch is used at first for startup. Then once suite2p or caiman is determined, the environment is switched.

milesAraya replied:
It is possible to remove expdb_batch however the method of initiating 

```bash
docker exec -it optinist-for-server-studio-dev-be-1 bash # Enter container 
conda activate expdb_batch # Activate conda environment 
alembic upgrade head # Run migrations 
echo "command: regist" > /app/experiments_datasets/M000000/M000000_ori001.proc 
# reset .proc file
python run_expdb_batch.py -o 1 # Run the batch process 
```

changes to 
```bash
docker exec -it optinist-for-server-studio-dev-be-1 bash # Enter container 
conda activate expdb_batch_caiman ### Activate conda environment (the correct one will be selected later)
alembic upgrade head # Run migrations 
echo "command: regist" > /app/experiments_datasets/M000000/M000000_ori001.proc 
# reset .proc file
python run_expdb_batch.py -o 1 # Run the batch process 
```

If you are happy with this, I can delete expdb_batch

itutu-tienday replied:
This seems to be related to the question raised in studio/app/optinist/core/expdb/batch_runner.py. 
(The purpose of activating conda in the python process)


After I have checked the design policy and the questions, I will comment again with my opinion.

milesAraya replied:
Removed. See https://github.com/arayabrain/optinist-for-server/pull/468#discussion_r2525560020

itutu-tienday replied:
I have confirmed the deletion of expdb_batch.

In addition, related adjustments are planned for the startup script (run_expdb_batch_script.sh) and batch_runner.py, so after checking and updating those, this post will be closed.

itutu-tienday replied:
Completed refactoring in #481

itutu-tienday commented on studio/app/optinist/core/expdb/batch_unit.py:
```diff
@@ -163,6 +168,229 @@ def check_is_circular_data(self, data_name: str):
         is_circular_data = any(pattern in data_name for pattern in circular_patterns)
         return is_circular_data
 
+    def _load_workflow_config(self):
```
Can't this function be replaced directly with `ExpDbValidator.validate_batch_nodes_in_workflow`?

itutu-tienday replied:
I understand the intention. I will consider organizing the code here.

milesAraya replied:
I think we would need to update ExpDbValidator.validate_batch_nodes_in_workflow to accept get_orimap, pca_analysis, kmeans_analysis.  Maybe it is best to create a new ExpDbValidator function?

itutu-tienday replied:
Please let me check the details separately (issues, expected specifications)

itutu-tienday commented on studio/app/optinist/core/expdb/batch_unit.py:
```diff
@@ -163,6 +168,229 @@ def check_is_circular_data(self, data_name: str):
+            )
+            return None
+
+    def _extract_workflow_param_values(self, workflow_params: dict) -> dict:
```
Are you aware that `WorkflowConfigReader.extract_node_names_in_workflow` is provided to perform this process?

milesAraya replied:
These do different things
  WorkflowConfigReader.extract_node_names_in_workflow(config)
  - Purpose: Extract node names from workflow
  - Returns: List[str] - e.g., ["preprocessing", "caiman_cnmf_preprocessing", "analyze_stats"]
  - Usage: Validation (which nodes exist in the workflow)

  _extract_workflow_param_values(workflow_params)
  - Purpose: Extract parameter values from nested workflow structure
  - Input: {param_name: {path, type, value}} or nested parent/children
  - Returns: {param_name: value} - flattened parameter dict
  - Usage: Get actual parameter values to pass to algorithms

But i will create WorkflowConfigReader.extract_workflow_param_values so that this is clearer and more organised.

itutu-tienday replied:
I understand the intention. I will consider organizing the code here.

itutu-tienday replied:
Addressed in #479

itutu-tienday commented on studio/app/optinist/core/expdb/batch_unit.py:
```diff
@@ -163,6 +168,229 @@ def check_is_circular_data(self, data_name: str):
+
+        from studio.app.expdb_dir_path import EXPDB_DIRPATH
+
+        subject_id = self.exp_id.split("_")[0]
```
Please use ExpDbPathIds for this parsing.

itutu-tienday commented on studio/app/optinist/core/expdb/batch_unit.py:
```diff
@@ -163,6 +168,229 @@ def check_is_circular_data(self, data_name: str):
+
+        return default_params
+
+    def _get_roi_method(self) -> str:
```
(Memo)
This process should not be defined within the batch_utit module, but should be defined using a separate proc file reader module.
*I plan to prepare the above reader module.

itutu-tienday replied:
The following PR provides reader/writer modules for proc files.
(These modules wrap operations on raw YAML files.)

- #473
-> ProcFileReader, ProcFileWriter

Once the PR is approved/merged, please use the reader/writer modules for proc file operations.

milesAraya replied:
Can this be considered resolved?

itutu-tienday replied:
It seems that `def _get_roi_method` itself will be built into ProcFileReader. It will be organized together with other functions. (I'll do it)

itutu-tienday replied:
Addressed in #479

itutu-tienday commented on studio/app/optinist/core/expdb/batch_runner.py:
```diff
@@ -269,6 +304,140 @@ class ExpDbBatchConcurrentProcess:
+
+        # Build command to run in specific conda environment
+        # We use conda run to execute Python in the target environment
+        cmd = [
```
The conda environment switching should have already been performed in advance in `cron/run_expdb_batch_script.sh`, but is switching the conda environment within the python process necessary?

If it's just a precautionary measure, I think it's best to avoid switching the conda environment within the python process if possible (to simplify the process).

Please let me know your thoughts.

milesAraya replied:
Okay, i have removed. Now for testing we need to explicitly set the correct conda env
`(expdb_batch_caiman) $ python run_expdb_batch.py -o 1 --filter-roi-method caiman`
or 
`(expdb_batch_suite2p) $ python run_expdb_batch.py -o 1 --filter-roi-method suite2p`

itutu-tienday replied:
Completed refactoring in #481

milesAraya committed:
Merge branch 'develop-suite2p-and-batch-gui' into feature/conda-env-switch

milesAraya committed:
Remove unused conda env, move conda switching to cron only, use pre-existing codebase over new functions

itutu-tienday commented on studio/app/optinist/core/expdb/batch_unit.py:
```diff
@@ -335,6 +558,83 @@ def cell_detection_cnmf(self, stack: ImageData):
+            return self.cell_detection_cnmf(stack)
+
     @stopwatch(callback=__stopwatch_callback)
     def generate_statdata(self) -> StatData:
```
The params passed to analyze_stats must also be obtained from workflow.yaml.

itutu-tienday commented on studio/app/optinist/core/expdb/batch_unit.py:
```diff
@@ -313,9 +529,16 @@ def generate_orimaps(self, stack: ImageData):
 
```
The parameters passed to preprocessing (call from def preprocess) must also come from workflow.yaml.

milesAraya replied:
I'm a but confused with the comment. _get_params_from_workflow_or_default does get read these. 
  1. If a workflow.yaml exists for the experiment AND contains a node named "preprocessing" → uses those params
  2. Otherwise → falls back to default params from preprocessing.yaml

  The same applies for "analyze_stats".

Have i misunderstood your point? Is there a specific case you are thinking about

itutu-tienday replied:
For example, isn't `get_default_params` hard-coded at the position of the following line number?
- https://github.com/arayabrain/optinist-for-server/blob/3836f5ee0df19d8d5ee85d1219403de70c844a94/studio/app/optinist/core/expdb/batch_unit.py#L501
  -  > params=get_default_params(“preprocessing”)

Since `_get_params_from_workflow_or_default` doesn't seem to be used in `preprocessing` or `analyze_stats`, I'm asking about this.
*If there still seems to be a misunderstanding between us, let's confirm it through conversation.

milesAraya committed:
Merge branch 'develop-suite2p-and-batch-gui' into feature/conda-env-switch

milesAraya committed:
Update to use _get_params_from_workflow_or_default always

milesAraya committed:
Merge branch 'develop-suite2p-and-batch-gui' into feature/conda-env-switch

Commit:
Merge branch 'develop-suite2p-and-batch-gui' into feature/conda-env-switch

Commit:
Added _read_from_any_path to WorkflowConfigReader

Commit:
suite2p preprocessing) refactoring.ExpDbBatch._load_workflow_config

Commit:
suite2p preprocessing) Added ExpDbWorkflowConfigReader and moved some methods from ExpDbBatch

Commit:
Added read_default_params function to workflow_params

Commit:
suite2p preprocessing) refactoring workflow params reading
- Removed WorkflowConfigReader.extract_workflow_param_values (duplicate of workflow_params.get_typecheck_params)

Commit:
params/suite2p_preprocessing.yaml is separated into development (lightweight version) and production

Commit:
suite2p preprocessing) Refactoring ProcFileReader.find_proc_files (WIP)

Commit:
suite2p preprocessing) Refactoring the passing of roi_method
- The loading points of proc file/roi_method are consolidated into one place (the entry point of ExpDbBatchConcurrentProcess).

itutu-tienday committed:
Merge pull request #480 from arayabrain/feature/suite2p-roi-detection-number-control

suite2p_preprocessing.yaml is separated into development and production

itutu-tienday committed:
Merge pull request #479 from arayabrain/feature/conda-env-switch-refactoring

Refactoring Feature/conda-env-switch branch (workflow params loading related)

Commit:
suite2p preprocessing) Refactoring: Simplify conda env switching process for roi_method

Commit:
suite2p preprocessing) Refactoring: Simplify conda env switching process for roi_method (launch .sh)

itutu-tienday committed:
Merge pull request #481 from arayabrain/feature/conda-env-switch-refactoring

Refactoring Feature/conda-env-switch branch (conda env and roi_method switching)

itutu-tienday commented on studio/app/dir_path.py:
```diff
@@ -67,3 +67,4 @@ class DIRPATH:
 class CORE_PARAM_PATH(Enum):
     nwb = f"{DIRPATH.APP_DIR}/optinist/core/nwb/nwb.yaml"
     snakemake = f"{DIRPATH.APP_DIR}/common/core/snakemake/snakemake.yaml"
+    batch_unit = f"{DIRPATH.APP_DIR}/optinist/core/expdb/params/batch_unit.yaml"
```
Is this property `batch_unit` currently in use?
If not, delete it.

milesAraya committed:
Remove unused batch_unit.yaml references

itutu-tienday commented:
@milesAraya 
This PR has been approved.
If there are no other actions remaining, please merge it.
